snapcraft/pluginhandler.py

```diff
@@ -109,21 +110,41 @@ def makedirs(self):
     def notify_stage(self, stage, hint=''):
         logger.info('%s %s %s', stage, self.name, hint)
 
-    def is_dirty(self, stage):
-        try:
-            with open(self.statefile, 'r') as f:
-                lastStep = f.read()
-                return (common.COMMAND_ORDER.index(stage) >
-                        common.COMMAND_ORDER.index(lastStep))
-        except Exception:
-            return True
+    def last_step(self):
+        for step in reversed(common.COMMAND_ORDER):
+            if os.path.exists(self._step_state_file(step)):
+                return step
+
+        return None
+
+    def is_dirty(self, step):
+        last_step = self.last_step()
+        if last_step:
+            return (common.COMMAND_ORDER.index(step) >
+                    common.COMMAND_ORDER.index(last_step))
+
+        return True
 
     def should_step_run(self, step, force=False):
         return force or self.is_dirty(step)
 
-    def mark_done(self, stage):
-        with open(self.statefile, 'w+') as f:
-            f.write(stage)
+    def mark_done(self, step):
+        index = common.COMMAND_ORDER.index(step)
+
+        # Currently only creating the step file. TODO: This will be a YAML file
+        # holding step-specific state information.
+        open(self._step_state_file(step), 'w').close()
+
+        # We know we've only just completed this step, so make sure any later
+        # steps don't have a saved state.
+        if index+1 != len(common.COMMAND_ORDER):
+            for command in common.COMMAND_ORDER[index+1:]:
+                state_file = self._step_state_file(command)
+                if os.path.exists(state_file):
+                    os.remove(state_file)
```

Comment:
Now that tests have past and I see coverage has dropped I'm going to be pedantic and say this line isn't tested :wink:

https://coveralls.io/builds/5471541/source?filename=snapcraft%2Fpluginhandler.py#L144

And it is true, I don't see a unit test that covers this case.